src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+import com.example.exception.CartItemAbsenceException;
+import com.example.exception.EntityNotFoundException;
+import com.example.mapper.CartItemMapper;
+import com.example.mapper.ShoppingCartMapper;
+import com.example.model.CartItem;
+import com.example.model.ShoppingCart;
+import com.example.repository.book.BookRepository;
+import com.example.repository.cartitem.CartItemRepository;
+import com.example.repository.shoppingcartrepository.ShoppingCartRepository;
+import com.example.service.ShoppingCartService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartResponseDto getByUserId(Long userId) {
+        return shoppingCartMapper.toDto(getShoppingCart(userId));
+    }
+
+    @Override
+    public ShoppingCartResponseDto addCartItem(Long userId, CreateCartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = getShoppingCart(userId);
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+        checkIfBookExist(requestDto);
+
+        if (shoppingCart.getCartItems().stream()
+                .anyMatch(item -> item.getBook().getId().equals(requestDto.getBookId()))) {
+            throw new BookAlreadyAddedToCartException("Book already added in shopping cart");
+        }
+
+        cartItem.setShoppingCart(shoppingCart);
+        CartItem save = cartItemRepository.save(cartItem);
+        CartItemResponseDto dto = cartItemMapper.toDto(save);
+        dto.setBookTitle(bookRepository.findById(dto.getBookId()).orElseThrow().getTitle());
+
+        ShoppingCartResponseDto resultDto = shoppingCartMapper.toDto(shoppingCart);
+        resultDto.getCartItems().add(dto);
+        return resultDto;
+    }
+
+    @Override
+    public ShoppingCartResponseDto updateCartItem(
+            Long userId,
+            Long cartItemId,
+            UpdateCartItemRequestDto requestDto
+    ) {
+        checkCartItemsInUsersShoppingCart(userId, cartItemId);
+        CartItem cartItemFromDb = cartItemRepository.findById(cartItemId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find cart item by id " + cartItemId));
+
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+        cartItemFromDb.setQuantity(cartItem.getQuantity());
+        cartItemRepository.save(cartItemFromDb);
+        return shoppingCartMapper.toDto(getShoppingCart(userId));
+    }
+
+    @Override
+    public ShoppingCartResponseDto deleteCartItem(Long userId, Long cartItemId) {
+        checkCartItemsInUsersShoppingCart(userId, cartItemId);
```

Comment:
Suggest other implementation

```suggestion
       ShoppingCart shoppingCart = repository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, shoppingCart.getId())
                .orElseThrow(() -> new EntityNotFoundException(
                        String.format("No such cartItem with id %d in shopping cart", cartItemId)
                ));
        shoppingCart.removeItemFromCart(cartItem);
```

Where in ShoppingCart class

```
public void removeItemFromCart(CartItem cartItem) {
        cartItems.remove(cartItem);
    }
```

Reply:
I'm not sure that correctly understand you, but option with

public void removeItemFromCart(CartItem cartItem) {
        cartItems.remove(cartItem);
    }
    
    in ShoppingCart class doesn't work for me, its actually not delete an item.
    
    So implement a little bit differ.

Reply:
@Elena-Bruyako thank you! refactored